packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
 /// <reference path="extra-types.d.ts" />
 
 /**
- * @typedef { "bigint" | "boolean" | "null" | "number" | "string" | "symbol" | "undefined" | "copyArray" | "copyRecord" | "copyError" | "promise" | "presence" } PassStyle
- * TODO "presence" above should indirect through REMOTE_STYLE to prepare
- * for changing it to "remotable"
+ * @typedef { "undefined" | "null" |
+ *   "boolean" | "number" | "bigint" | "string" | "symbol" |
+ *   "copyArray" | "copyRecord" | "copyError" | "remotable" | "promise"
```

Comment:
BTW, there's some amount of [reference-mechanicness](http://erights.org/elib/concurrency/refmech.html#legend) that happens in Goblins... remote references, when sent across the wire, and then when rehydrated on the other end, are tagged as either references to remote objects or remote promises.  So we basically have four kinds of live-refs:

- local-object
- local-promise
- remote-object
- remote-promise

(Or see [this table](https://gitlab.com/spritely/goblins/-/blob/4f664218c87fa96432a04d0a107618ec36ab9310/goblins/core.rkt#L345).)

The reason for this is knowing whether or not any further state transitions may happen for pipelining purposes.  I would have to look carefully to remember the exact details of why these are kept distinct but I remember it was related to pipelining.  (Whether or not a promise has resolved is not carried across the wire, however.)

I don't think this is urgent to resolve right now, but I wanted to bring it up, just in case it's relevant to the future.  IIUC that's not marked distinctly right now I think in the JS side.

Reply:
I thought I'd provide more information about why this information is tagged as such.  This is not essential to @erights' writeup, I just figured I'd help explain why it was important to Goblins' design.

In Goblins, refs are actually keys compared by their `eq?` (opaque pointer) identity in a (hash)map called an "actormap", more or less the heap of a vat.  At its simplest, synchronous early implementation, this was merely a mapping of `{ref: message-handler-procedure}`.  However, as time has gone on, the behavior of objects underlying the conceptual idea of an object has had to become more rich.  As such, the "mactor" was introduced, so that the mapping became `{ref: mactor}`.  There are a [number of types of mactors](https://gitlab.com/spritely/goblins/-/blob/4f664218c87fa96432a04d0a107618ec36ab9310/goblins/core.rkt#L475), with the simple handler-procedure version being `mactor:object` (a struct that contains the current message-handler-procedure).  But there are also other kinds, such as `mactor:naive` and `mactor:question` (the two starting positions of promises, the former being the default type, the latter being ones which are, well, questions from a captp kind of perspective).  Here is the transition state diagram, copied from the code:

```

;;;        Unresolved                     Resolved
;;;  __________________________  ___________________________
;;; |                          ||                           |
;;;
;;;                 .----------------->.    .-->[object]
;;;                 |                  |    |
;;;                 |    .--.          |    +-->[local-link]
;;;     [naive]-->. |    v  |          |    |            
;;;               +>+->[closer]------->'--->+-->[encased]
;;;  [question]-->' |       |               |            
;;;                 |       |               '-->[broken]
;;;                 '------>'--->[remote-link]    ^
;;;                                  |            |
;;;                                  '----------->'
;;;
;;; |________________________________________||_____________|
;;;                  Eventual                     Settled
```

A naive/question promise will transition states as follows:

- `mactor:closer` if it points to another promise whether local or remote (shortening as far as it can figure out how to do, while maintaining a set of past transitions, so as to detect a cycle)
- `mactor:remote-link` if resolving to a remote object (as opposed to remote promise!)
- `mactor:local-link` if it has settled on a local object (if near, we can now even use `$` for synchronous operations as if calling whatever it links to)
- `mactor:encased` if it settles to a data value (in which case it has to be extracted with `on`, or opportunistically can be extracted with `$` if the promise is near)
- `mactor:broken` if the promise breaks, for whatever reason, including (in the future, not supported yet) if a corresponding connection over CapTP has broken.

I realize this is non-critical to the work @erights is doing in this commit, but I wanted to explain this to get the idea out there about Goblins' concept of transition states, and how processing an incoming promise resolution occurs, and how in particular how the typing information for either a promise or an object determines whether the actormap will transition the ref to either a `mactor:remote-link` or `mactor:closer` depending on whether or not the remote side is tagged as a remote object or a remote promise.

(It may be that there is a more obvious way to do things too without sharing this information which I have not yet seen.)